### ISSUE TITLE
Confusion on Using TimeDistributedLayer on DynamicRNNLayer

### ISSUE THREAD

**Issue by Zsank** [zsdonghao/tensorlayer#112]
Hi! I want to deal with sentences with different length and I had tried to used TimeDistributedLayer on DynamicRNNLayer. My code was:
```
encode_seqs = tf.placeholder(tf.int64, shape=[5, None], name='encode_seqs')
en_net = tl.layers.EmbeddingInputlayer(inputs=encode_seqs,
                                       vocabulary_size=vocab_size,
                                       embedding_size=hidden_size,
                                       E_init=tf.random_uniform_initializer(-init_scale, init_scale),
                                       name="encode_seqs")
network = tl.layers.TimeDistributedLayer(layer=en_net,
                                         layer_class=tl.layers.DynamicRNNLayer,
                                         args={'cell_fn': tf.contrib.rnn.BasicLSTMCell,
                                               'n_hidden': hidden_size,
                                               'initializer':
                                                    tf.random_uniform_initializer(-init_scale, init_scale),
                                                'sequence_length':
                                                    tl.layers.retrieve_seq_length_op2(encode_seqs),
                                                'initial_state': True,
                                                'dropout': dropout,
                                                'n_layer': 1,
                                                'return_last': True,
                                                'name': 'dynamic_rnn'},
                                          name='encode_TokensToSeqs_level')
 network = tl.layers.DynamicRNNLayer(layer=network,
                                     cell_fn=tf.contrib.rnn.BasicLSTMCell,
                                     cell_init_args={'state_is_tuple': False},
                                     n_hidden=hidden_size,
                                     initializer=tf.random_uniform_initializer(-init_scale, init_scale),
                                     sequence_length=tl.layers.retrieve_seq_length_op2(encode_seqs),
                                     dropout=dropout,
                                     n_layer=1,
                                     return_last=True,
                                     name='encode_SeqsToText_level')
```
and the error is "ValueError: Cannot infer num from shape (5, ?, 200)"
(The embedding_size I set was 200)

In addition, when I replaced None with specific number such as 30, then the error became "ValueError: shape(5, 200) and (?,?,?) should have the same rank." I did not know why.

**Comment by zsdonghao** [comment]
if your input is a sentence, just use `DynamicRNNLayer`.

the example from [DynamicRNNLayer](http://tensorlayer.readthedocs.io/en/latest/modules/layers.html#id26) should work in your case.

**Comment by Zsank** [comment]
Sure that the example from DynamicRNNLayer can work. However, what I want is to input more than one sentences and apply DynamicRNNLayer on each of them simultaneously. Then use their outputs as input to the next DynamicRNNLayer, just like my defined name in layers "TokensToSeqs" "Seq2ToText". So, if TimeDistributedLayer can be used on DynamicRNNLayer? Or should I use a loop for it?

**Comment by zsdonghao** [comment]
@Zsank as you set `return_last=True`, you can encode a batch of sentences simultaneously, but what is the other `DynamicRNNLayer` for?

I think `TimeDistributedLayer` have nothing to do in your case, because it is for non-text input.

**Comment by Zsank** [comment]
@zsdonghao I am sorry that maybe I had not describe my confusion clearly. 
Before `TimeDistributedLayer` I use some functions from `tl.nlp` to transfer words-in-sentences to integer index and  `EmbeddingInputLayer` to become embedded word vector, so the inputs to `TimeDistributedLayer` is not text input but vector input with `[batch_size, n_steps, n_dim]`. So I think `TimeDistributedLayer` maybe helpful. But the problem is that `ValueError: Cannot infer num from shape (5, ?, 200)` and I didn't know why.
As for `DynamicRNNLayer`, the first one I used as a `layer_class` in `TimeDistributedLayer` in order to encode at the sentence_level (That means encode tokens into sentences). The other `DynamicRNNLayer` I used was for encoding at the document_level (That means encode sentences into paragraph/document).
